File: modules/wem-api/src/main/java/com/enonic/wem/api/module/ResourcePath.java
package com.enonic.wem.api.module;


import com.google.common.base.Splitter;
import com.google.common.collect.ImmutableList;

import com.enonic.wem.api.BasePath;

public final class ResourcePath
    extends BasePath<ResourcePath>
{
    private static final char SEPARATOR = '/';

    private static final ResourcePath ROOT = ResourcePath.from( "/" );

    private ResourcePath( final Iterable<String> elements, final boolean isAbsolute )
    {
        super( SEPARATOR, elements, isAbsolute );
    }

    @Override
    protected ResourcePath newPath( final Iterable<String> elements, final boolean isAbsolute )
    {
        return new ResourcePath( elements, isAbsolute );
    }

    public static ResourcePath root()
    {
        return ROOT;
    }

    public static ResourcePath from( final String path )
    {
        final Iterable<String> pathElements = Splitter.on( SEPARATOR ).omitEmptyStrings().split( path );
        final boolean isAbsolute = path.startsWith( String.valueOf( SEPARATOR ) );
        return new ResourcePath( pathElements, isAbsolute );
    }

    public static ResourcePath from( final ResourcePath path, final String name )
    {
        ImmutableList.Builder<String> builder = new ImmutableList.Builder<>();
        builder.addAll( path );
        builder.add( name );

        return new ResourcePath( builder.build(), path.isAbsolute() );
    }

}

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/application/ResourcePath.ts
module api.application {

    export class ResourcePath extends api.util.BasePath<ResourcePath> {

        private static ELEMENT_DIVIDER:string = "/";

        public static fromString(s:string) {
            var absolute:boolean = s.charAt(0) == ResourcePath.ELEMENT_DIVIDER;
            var elements:string[] = s.split(ResourcePath.ELEMENT_DIVIDER);
            return new ResourcePath(api.util.BasePath.removeEmptyElements(elements), absolute);
        }

        constructor(elements:string[], absolute?:boolean) {
            super(elements, ResourcePath.ELEMENT_DIVIDER, absolute);
        }

        newInstance(elements:string[], absolute:boolean):ResourcePath {
            return new ResourcePath(elements, absolute);
        }
    }
}